ConwaysGameOfLife/GameController.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+
+namespace ConwaysGameOfLife
+{
+    public class GameController
+    {
+        // responsible for implementing game
+        // orchestrator type class - brings all together
+
+        /*
+            welcome message
+            user asked for universe dimensions
+            dimensions parsed
+            universe initialized
+            initial universe printed to console
+            user asked for location of alive cells
+            input parsed/validated
+            universe set up with live cells
+            universe printed
+            game plays
+            rules checked
+            new universe regenerated
+            ends either by user typing q or all dead cells
+
+            input
+            output
+            universe
+            rules
+        */
+        private Universe _universe;
+
+        private IOutput _output;
+
+        private IInput _input;
+
+        public GameController(Universe universe, IOutput output, IInput input)
+        {
+            _universe = universe;
+            _output = output;
+            _input = input;
+        }
+
+        public void CreateInitialUniverse()
+        {
+            _output.WriteLine(Messages.RequestDimensions);
+            var input = CreateValidUniverseString();
+            InputParser.ParseUniverse(input);
+            DisplayUniverse();
+        }
+
+        public void PopulateUniverseWithLiveCells()
+        {
+            // _output.Write(Messages.RequestLiveCell);
+            // _output.WriteLine($"Or {Messages.Play});
+            // loop
+            // var locationInput = _input.ReadLine();
+            // var location = SetLiveCellLocation(locationInput, universe.GridWidth, universe.GridLength);
+            // universe.SwitchCellState(universe.GetCellAtLocation(location.X, location.Y));
+            // DisplayUniverse();
+            throw new NotImplementedException();
+        }
+
+        public void DisplayUniverse() 
+        {
+            _output.WriteLine(OutputFormatter.FormatUniverse(_universe));
+        }
+
+        public void Run()
+        {
+            _output.WriteLine(Messages.Welcome);
+            // create initial universe
+            CreateInitialUniverse();
+            // add live cells to universe until user presses 'p' to play
+            // generator checks all cells if change of state required & creates next universe
+            // loop last step until user presses 'q' to quit or all cells are dead
+        }
+        
+        public string CreateValidUniverseString()
+        {
+            var input = _input.ReadLine();
+            var isValidUniverse = Validator.IsValidUniverse(input);
+            while (!isValidUniverse)
+            {
+                InvalidUniverseString(input);
+            }
+            return input;
+        }
+
+        private void InvalidUniverseString(string input)

```

Comment:
I think it's better if you just put the code within this function directly in the while loop as it's doing 2 different things

Reply:
Generally wouldn't comment for resource optimisation, but in this case, you'd be unnecessarily creating recursive calls to `CreateValidUniverseString` whenever the input is invalid. This could, however unlikely, lead to a memory problem as you're storing the `input` and `isValidUniverse` values for each recursive call until you finally get valid input. Here, another problem is presented because we're not only making recursive calls, but we're also using a while loop. The while loop will loop forever because we never actually save anything inside and reevaluate the predicate!

For demonstration:
1. `CreateValidUniverseString` is called
2. `input` is assigned some invalid input `"invalid1"`
3. `isValidUniverse` is evaluated to false
4. `!isValidUniverse` is evaluiated to true, we go into the while loop
5. `InvalidUniverseString("invalid1")` is called
6. We write to the output to request dimensions again
7. `CreateValidUniverseString` is called again, repeating steps 1-6 but potentially with any invalid input; until
8. `CreateValidUniverseString` is called for the final time
9. `input`  is assigned some valid input `"3,3"`
10. `isValidUniverse` is evaluated to true
11. `!isValidUniverse` is evaluated to false
12. `input` is returned to the latest `CreateValidUniverseString` call from 7.
13. We're now back at the end of the latest `CreateValidUniverseString` call's while loop
14. `!isValidUniverse` gets evaluated to true (!!!) because the `input` from 12. isn't stored and reevaluated
15. `InvalidUniverseString` is called again 
16. We're just stuck in a cycle even if the user repeatedly puts valid input

Recommendations:
- in general, choose only one of recursion or while loop for any given function
  - for this particular function, recursion gives no benefit (we can assign some reading if you're interested in recursion and where it might be useful)
- as @indianbread suggested, move the `InvalidUniverseString` lines 91 and 92 directly inside the while loop of the `CreateValidUniversString`. This should help clarify the demonstration above.
- once you understand the demonstration:
  1. change the code from line 92 to just read a new line in like on line 80
  2. add another line to reevaluate `isValidUniverse` with the new input
  3. let the while loop do the hard work until some valid input is reached! 🎉